File: lib/features/cams/data/services/store_hub_service.dart
import 'dart:async';
import 'package:flutter/foundation.dart';
import 'package:signalr_netcore/signalr_client.dart';
import '../../../../core/constants/api_constants.dart';
import '../../../../core/enums/playback_command_enum.dart';
import '../../../../core/enums/transition_type_enum.dart';
import '../../../suno/domain/entities/suno_generation_status.dart';
import '../models/space_playback_state_model.dart';

/// Service for real-time communication with CAMS StoreHub via SignalR.
///
/// Events received (Server → Client):
/// - `PlayStream`           — AI/override changed the playlist
/// - `PlaybackStateChanged` — Playback command relay (Pause/Resume/Seek/Skip)
/// - `SpaceStateSync`       — Full state re-sync after cancel override
/// - `StopPlayback`         — Stop all playback
/// - `ConnectionConfirmed`  — Acknowledgement after joining a Space/ManagerRoom
///
/// Methods sent (Client → Server):
/// - `JoinSpaceAsync`       — Tablet joins Space group
/// - `LeaveSpaceAsync`      — Tablet leaves Space group
/// - `ReportPlaybackStateAsync` — Health/analytics reporting
class StoreHubService {
  final String Function() _accessTokenFactory;
  HubConnection? _connection;
  String? _currentSpaceId;
  String? _currentManagerStoreId;
  String? _currentManagerBrandId;

  /// Offset in milliseconds: (deviceTimeUtc - serverTimeUtc).
  /// Positive = device clock is ahead of server.
  /// Used to compensate seek calculations derived from `startedAtUtc`.
  int _serverClockOffsetMs = 0;

  /// Public getter for the calculated server clock offset.
  int get serverClockOffsetMs => _serverClockOffsetMs;

  StoreHubService({required String Function() accessTokenFactory})
      : _accessTokenFactory = accessTokenFactory;

  // ─── Stream controllers for broadcasting events ─────────────────────────

  final _playStreamController = StreamController<PlayStreamEvent>.broadcast();
  final _playbackCommandController =
      StreamController<PlaybackCommandEvent>.broadcast();
  final _statesSyncController =
      StreamController<SpacePlaybackStateModel>.broadcast();
  final _stopPlaybackController = StreamController<void>.broadcast();
  final _sunoGenerationStatusController =
      StreamController<SunoGenerationStatusChangedEvent>.broadcast();
  final _connectionController = StreamController<ConnectionStatus>.broadcast();

  /// Stream of PlayStream events (playlist changes).
  Stream<PlayStreamEvent> get onPlayStream => _playStreamController.stream;

  /// Stream of playback commands from manager.
  Stream<PlaybackCommandEvent> get onPlaybackCommand =>
      _playbackCommandController.stream;

  /// Stream of full state syncs (after cancel override).
  Stream<SpacePlaybackStateModel> get onSpaceStateSync =>
      _statesSyncController.stream;

  /// Stream of stop playback events.
  Stream<void> get onStopPlayback => _stopPlaybackController.stream;

  /// Stream of Suno async generation status changes.
  Stream<SunoGenerationStatusChangedEvent> get onSunoGenerationStatusChanged =>
      _sunoGenerationStatusController.stream;

  /// Stream of connection status changes.
  Stream<ConnectionStatus> get onConnectionStatus =>
      _connectionController.stream;

  /// Current connection state.
  bool get isConnected => _connection?.state == HubConnectionState.Connected;

  // ─── Lifecycle ───────────────────────────────────────────────────────────

  /// Connect to the StoreHub.
  Future<void> connect() async {
    if (_connection != null &&
        _connection!.state == HubConnectionState.Connected) {
      return; // Already connected
    }

    _connection = HubConnectionBuilder()
        .withUrl(
      ApiConstants.storeHubUrl,
      options: HttpConnectionOptions(
        accessTokenFactory: () async => _accessTokenFactory(),
        transport: HttpTransportType.WebSockets,
        skipNegotiation: true,
      ),
    )
        .withAutomaticReconnect(
      retryDelays: [0, 2000, 5000, 10000, 30000],
    ).build();

    _registerListeners();
    await _connection!.start();

    _connectionController.add(ConnectionStatus.connected);
  }

  /// Disconnect from the StoreHub.
  Future<void> disconnect() async {
    if (_currentSpaceId != null) {
      await leaveSpace(_currentSpaceId!);
    }
    if (_currentManagerStoreId != null) {
      await leaveManagerRoom(_currentManagerStoreId!);
    }
    if (_currentManagerBrandId != null) {
      await leaveBrandManagerRoom(_currentManagerBrandId!);
    }
    await _connection?.stop();
    _connection = null;
    _connectionController.add(ConnectionStatus.disconnected);
  }

  // ─── Client → Server methods ─────────────────────────────────────────────

  /// Join a Space group to receive its events.
  Future<void> joinSpace(String spaceId) async {
    _currentSpaceId = spaceId;
    await _connection?.invoke('JoinSpaceAsync', args: [spaceId]);
  }

  /// Leave a Space group.
  Future<void> leaveSpace(String spaceId) async {
    await _connection?.invoke('LeaveSpaceAsync', args: [spaceId]);
    if (_currentSpaceId == spaceId) {
      _currentSpaceId = null;
    }
  }

  Future<void> joinManagerRoom(String storeId) async {
    _currentManagerStoreId = storeId;
    await _connection?.invoke('JoinManagerRoomAsync', args: [storeId]);
  }

  Future<void> leaveManagerRoom(String storeId) async {
    try {
      await _connection?.invoke('LeaveManagerRoomAsync', args: [storeId]);
    } catch (_) {
      // Some environments may not expose LeaveManagerRoomAsync yet.
    }
    if (_currentManagerStoreId == storeId) {
      _currentManagerStoreId = null;
    }
  }

  Future<void> joinBrandManagerRoom(String brandId) async {
    _currentManagerBrandId = brandId;
    await _connection?.invoke('JoinBrandManagerRoomAsync', args: [brandId]);
  }

  Future<void> leaveBrandManagerRoom(String brandId) async {
    try {
      await _connection?.invoke('LeaveBrandManagerRoomAsync', args: [brandId]);
    } catch (_) {
      // Some environments may not expose LeaveBrandManagerRoomAsync yet.
    }
    if (_currentManagerBrandId == brandId) {
      _currentManagerBrandId = null;
    }
  }

  /// Report playback state for analytics/health monitoring.
  Future<void> reportPlaybackState({
    required String spaceId,
    required bool isPlaying,
    double? positionSeconds,
    String? currentHlsUrl,
  }) async {
    await _connection?.invoke('ReportPlaybackStateAsync', args: [
      {
        'spaceId': spaceId,
        'isPlaying': isPlaying,
        'positionSeconds': positionSeconds,
        'currentHlsUrl': currentHlsUrl,
      }
    ]);
  }

  // ─── Server → Client event listeners ──────────────────────────────────────

  void _registerListeners() {
    final conn = _connection;
    if (conn == null) return;

    // Connection confirmed after joining
    conn.on('ConnectionConfirmed', (args) {
      final data = _asMap(args?[0]);
      if (data != null) {
        debugPrint(
          '[StoreHub] Connected to group: ${_readString(data, 'spaceId') ?? _readString(data, 'storeId')}',
        );
        // Compute clock drift from server timestamp.
        final serverTimeUtc = _parseDateTime(_readValue(data, 'serverTimeUtc'));
        if (serverTimeUtc != null) {
          _serverClockOffsetMs =
              DateTime.now().toUtc().difference(serverTimeUtc.toUtc()).inMilliseconds;
          debugPrint(
            '[StoreHub] Clock offset: ${_serverClockOffsetMs}ms '
            '(positive = device ahead)',
          );
        }
      }
    });

    // Playlist changed (AI or override)
    conn.on('PlayStream', (args) {
      final payload = _asMap(args?[0]);
      if (payload == null) return;

      final transitionType =
          TransitionTypeEnum.fromJson(_readValue(payload, 'transitionType')) ??
              TransitionTypeEnum.immediate;
      final hlsUrl = _readString(payload, 'hlsUrl') ?? '';

      // Pending means stream is not ready yet.
      if (transitionType == TransitionTypeEnum.pending && hlsUrl.isEmpty) {
        return;
      }

      _playStreamController.add(PlayStreamEvent(
        spaceId: _readString(payload, 'spaceId') ?? '',
        hlsUrl: hlsUrl,
        playlistId: _readString(payload, 'playlistId'),
        currentQueueItemId: _readString(payload, 'currentQueueItemId'),
        trackId: _readString(payload, 'trackId'),
        trackName: _readString(payload, 'trackName'),
        transitionType: transitionType,
        isManualOverride: _readBool(payload, 'isManualOverride') ?? false,
        startedAtUtc: _parseDateTime(_readValue(payload, 'startedAtUtc')),
      ));
    });

    // Playback command from manager
    conn.on('PlaybackStateChanged', (args) {
      final payload = _asMap(args?[0]);
      if (payload == null) return;

      _playbackCommandController.add(PlaybackCommandEvent(
        spaceId: _readString(payload, 'spaceId') ?? '',
        command: PlaybackCommandEnum.fromValue(
          _readNum(payload, 'command')?.toInt() ?? 1,
        ),
        seekPositionSeconds:
            _readNum(payload, 'seekPositionSeconds')?.toDouble(),
        targetTrackId: _readString(payload, 'targetTrackId'),
      ));
    });

    // Full state sync (after cancel override)
    conn.on('SpaceStateSync', (args) {
      final payload = _asMap(args?[0]);
      if (payload != null) {
        _statesSyncController.add(
          SpacePlaybackStateModel.fromSignalR(_normalizeKeyCasing(payload)),
        );
      }
    });

    // Stop playback
    conn.on('StopPlayback', (_) {
      _stopPlaybackController.add(null);
    });

    conn.on('SunoGenerationStatusChanged', (args) {
      final payload = _asMap(args?[0]);
      if (payload == null) return;

      _sunoGenerationStatusController.add(
        SunoGenerationStatusChangedEvent(
          id: _readString(payload, 'id') ?? '',
          brandId: _readString(payload, 'brandId') ?? '',
          generationStatus:
              SunoGenerationStatus.fromJson(
                _readValue(payload, 'generationStatus'),
              ),
          progressPercent: _readNum(payload, 'progressPercent')?.toInt(),
          errorMessage: _readString(payload, 'errorMessage'),
          generatedTrackId: _readString(payload, 'generatedTrackId'),
        ),
      );
    });

    // Error from hub
    conn.on('Error', (args) {
      final message = args?[0] as String?;
      debugPrint('[StoreHub] Server error: $message');
    });

    // Reconnect lifecycle
    conn.onreconnecting(({error}) {
      debugPrint('[StoreHub] Reconnecting... $error');
      _connectionController.add(ConnectionStatus.reconnecting);
    });

    conn.onreconnected(({connectionId}) {
      debugPrint('[StoreHub] Reconnected: $connectionId');
      _connectionController.add(ConnectionStatus.connected);
      // Re-join Space after reconnect
      if (_currentSpaceId != null) {
        unawaited(joinSpace(_currentSpaceId!));
      }
      if (_currentManagerStoreId != null) {
        unawaited(joinManagerRoom(_currentManagerStoreId!));
      }
      if (_currentManagerBrandId != null) {
        unawaited(joinBrandManagerRoom(_currentManagerBrandId!));
      }
    });

    conn.onclose(({error}) {
      debugPrint('[StoreHub] Connection closed: $error');
      _connectionController.add(ConnectionStatus.disconnected);
    });
  }

  // ─── Cleanup ──────────────────────────────────────────────────────────────

  void dispose() {
    _connection?.stop();
    _playStreamController.close();
    _playbackCommandController.close();
    _statesSyncController.close();
    _stopPlaybackController.close();
    _sunoGenerationStatusController.close();
    _connectionController.close();
  }

  Map<String, dynamic>? _asMap(dynamic raw) {
    if (raw is Map<String, dynamic>) return raw;
    if (raw is Map) return Map<String, dynamic>.from(raw);
    return null;
  }

  Map<String, dynamic> _normalizeKeyCasing(Map<String, dynamic> payload) {
    final normalized = <String, dynamic>{};
    payload.forEach((key, value) {
      normalized[key] = value;
      if (key.isEmpty) return;
      final camelCaseKey = '${key[0].toLowerCase()}${key.substring(1)}';
      normalized.putIfAbsent(camelCaseKey, () => value);
    });
    return normalized;
  }

  dynamic _readValue(Map<String, dynamic> payload, String key) {
    if (payload.containsKey(key)) return payload[key];
    if (key.isEmpty) return null;
    final pascalCaseKey = '${key[0].toUpperCase()}${key.substring(1)}';
    if (payload.containsKey(pascalCaseKey)) return payload[pascalCaseKey];
    return null;
  }

  String? _readString(Map<String, dynamic> payload, String key) {
    final value = _readValue(payload, key);
    if (value == null) return null;
    if (value is String) return value;
    return value.toString();
  }

  num? _readNum(Map<String, dynamic> payload, String key) {
    final value = _readValue(payload, key);
    if (value is num) return value;
    if (value is String) return num.tryParse(value);
    return null;
  }

  bool? _readBool(Map<String, dynamic> payload, String key) {
    final value = _readValue(payload, key);
    if (value is bool) return value;
    if (value is String) {
      if (value.toLowerCase() == 'true') return true;
      if (value.toLowerCase() == 'false') return false;
    }
    return null;
  }

  DateTime? _parseDateTime(dynamic raw) {
    if (raw is DateTime) return raw;
    if (raw is String) return DateTime.tryParse(raw);
    return null;
  }
}

// ─── Event DTOs ───────────────────────────────────────────────────────────────

/// Payload for PlayStream event.
class PlayStreamEvent {
  final String spaceId;
  final String hlsUrl;
  final String? playlistId;
  final String? currentQueueItemId;
  final String? trackId;
  final String? trackName;
  final TransitionTypeEnum transitionType;
  final bool isManualOverride;
  final DateTime? startedAtUtc;

  const PlayStreamEvent({
    required this.spaceId,
    required this.hlsUrl,
    this.playlistId,
    this.currentQueueItemId,
    this.trackId,
    this.trackName,
    required this.transitionType,
    this.isManualOverride = false,
    this.startedAtUtc,
  });
}

/// Payload for PlaybackStateChanged event.
class PlaybackCommandEvent {
  final String spaceId;
  final PlaybackCommandEnum command;
  final double? seekPositionSeconds;
  final String? targetTrackId;

  const PlaybackCommandEvent({
    required this.spaceId,
    required this.command,
    this.seekPositionSeconds,
    this.targetTrackId,
  });
}

class SunoGenerationStatusChangedEvent {
  final String id;
  final String brandId;
  final SunoGenerationStatus generationStatus;
  final int? progressPercent;
  final String? errorMessage;
  final String? generatedTrackId;

  const SunoGenerationStatusChangedEvent({
    required this.id,
    required this.brandId,
    required this.generationStatus,
    this.progressPercent,
    this.errorMessage,
    this.generatedTrackId,
  });
}

/// Connection status enum for UI.
enum ConnectionStatus {
  connected,
  disconnected,
  reconnecting,
}

File: src/shared/modules/cams/services/storeHubService.ts
import {
  HubConnection,
  HubConnectionBuilder,
  LogLevel,
  HubConnectionState,
} from '@microsoft/signalr';
import { env } from '@/config';
import type {
  PlayStreamPayload,
  PlaybackStateChangedPayload,
  SpaceStateDto,
} from '../types';

type StoreHubEventHandlers = {
  onPlayStream?: (payload: PlayStreamPayload) => void;
  onPlaybackStateChanged?: (payload: PlaybackStateChangedPayload) => void;
  onSpaceStateSync?: (spaceId: string, state: SpaceStateDto) => void;
  onConnected?: () => void;
  onDisconnected?: () => void;
  onReconnecting?: () => void;
  onReconnected?: () => void;
};

class StoreHubService {
  private connection: HubConnection | null = null;
  private currentStoreId: string | null = null;
  private joinedSpaceIds = new Set<string>();
  private eventHandlers: StoreHubEventHandlers = {};

  /**
   * Clock offset in ms: clientNow - serverNow at connection time.
   * Positive means client clock is ahead of server.
   * Used by getEffectiveSeekOffset to correct seek positions.
   */
  private _serverClockOffsetMs = 0;

  /** Exposed to playbackHelpers so seek math can compensate for clock skew. */
  public get serverClockOffsetMs(): number {
    return this._serverClockOffsetMs;
  }

  /**
   * Check if connection is active
   */
  public isConnected(): boolean {
    return (
      this.connection !== null &&
      this.connection.state === HubConnectionState.Connected
    );
  }

  /**
   * Connect to StoreHub and join manager room
   */
  public async connect(
    storeId: string,
    token: string,
    handlers: StoreHubEventHandlers = {},
  ): Promise<void> {
    // ✅ Prevent double-connect
    if (this.isConnected() && this.currentStoreId === storeId) {
      console.log('⏭️ Already connected to this store');
      return;
    }

    try {
      // ✅ Disconnect existing connection first
      if (this.connection) {
        console.log('🔄 Disconnecting existing connection...');
        await this.disconnect();
      }

      this.eventHandlers = handlers;
      this.currentStoreId = storeId;

      const hubUrl = `${env.baseUrl}/hubs/store`;

      console.log('🔌 Connecting to StoreHub:', {
        baseUrl: env.baseUrl,
        hubUrl,
        storeId,
        hasToken: !!token,
      });

      // Build connection
      this.connection = new HubConnectionBuilder()
        .withUrl(hubUrl, {
          accessTokenFactory: () => token,
        })
        .withAutomaticReconnect({
          nextRetryDelayInMilliseconds: (retryContext) => {
            // Exponential backoff: 2s, 4s, 8s, 16s, then 30s max
            const delay = Math.min(
              2000 * Math.pow(2, retryContext.previousRetryCount),
              30000,
            );
            console.log(
              `⏰ Retry ${retryContext.previousRetryCount + 1} in ${delay}ms`,
            );
            return delay;
          },
        })
        .configureLogging(LogLevel.Information)
        .build();

      // Register event listeners
      this.registerEventListeners();

      // Start connection
      await this.connection.start();

      console.log('✅ StoreHub connected successfully');

      // ✅ Wait a bit before joining to ensure connection is stable
      await new Promise((resolve) => setTimeout(resolve, 100));

      // Join manager room
      await this.joinStore(storeId);

      // Trigger connected callback
      this.eventHandlers.onConnected?.();
    } catch (error) {
      console.error('❌ Failed to connect to StoreHub:', error);
      this.connection = null;
      this.currentStoreId = null;
      throw error;
    }
  }

  /**
   * Join manager room for a store
   */
  private async joinStore(storeId: string): Promise<void> {
    if (!this.connection) {
      throw new Error('Connection not initialized');
    }

    // ✅ Check connection state before invoking
    if (this.connection.state !== HubConnectionState.Connected) {
      throw new Error(
        `Cannot join store: Connection state is ${this.connection.state}`,
      );
    }

    try {
      console.log('🚪 Joining manager room for store:', storeId);
      await this.connection.invoke('JoinManagerRoomAsync', storeId);
      console.log('✅ Joined manager room successfully');
    } catch (error) {
      console.error('❌ Failed to join manager room:', error);
      throw error;
    }
  }

  /**
   * Leave current store's manager room
   */
  private async leaveStore(): Promise<void> {
    if (!this.connection || !this.currentStoreId) {
      return;
    }

    // ✅ Only try to leave if connected
    if (this.connection.state !== HubConnectionState.Connected) {
      console.log('⏭️ Connection not active, skipping leave');
      return;
    }

    try {
      console.log('🚪 Leaving manager room for store:', this.currentStoreId);
      await this.connection.invoke(
        'LeaveManagerRoomAsync',
        this.currentStoreId,
      );
      console.log('✅ Left manager room successfully');
    } catch (error) {
      console.error('❌ Failed to leave store:', error);
      // Don't throw - just log
    }
  }

  /**
   * Join a specific space group to receive real-time updates
   */
  public async joinSpace(spaceId: string): Promise<void> {
    if (!this.connection) {
      throw new Error('Connection not initialized');
    }

    if (this.connection.state !== HubConnectionState.Connected) {
      throw new Error(
        `Cannot join space: Connection state is ${this.connection.state}`,
      );
    }

    try {
      console.log('🎵 Joining space group:', spaceId);
      await this.connection.invoke('JoinSpaceAsync', spaceId);
      this.joinedSpaceIds.add(spaceId);
      console.log('✅ Joined space group successfully:', spaceId);
    } catch (error) {
      console.error('❌ Failed to join space group:', error);
      throw error;
    }
  }

  /**
   * Leave a specific space group
   */
  public async leaveSpace(spaceId: string): Promise<void> {
    if (!this.connection) {
      return;
    }

    if (this.connection.state !== HubConnectionState.Connected) {
      console.log('⏭️ Connection not active, skipping leave space');
      return;
    }

    try {
      console.log('👋 Leaving space group:', spaceId);
      await this.connection.invoke('LeaveSpaceAsync', spaceId);
      this.joinedSpaceIds.delete(spaceId);
      console.log('✅ Left space group successfully:', spaceId);
    } catch (error) {
      console.error('❌ Failed to leave space group:', error);
      // Don't throw - just log
    }
  }

  /**
   * Register SignalR event listeners
   */
  private registerEventListeners(): void {
    if (!this.connection) return;

    // Server ack after JoinSpaceAsync / JoinManagerRoomAsync (camelCase over the wire)
    this.connection.on(
      'connectionconfirmed',
      (data: { serverTimeUtc?: string } | null) => {
        if (data?.serverTimeUtc) {
          const serverMs = new Date(data.serverTimeUtc).getTime();
          this._serverClockOffsetMs = Date.now() - serverMs;
          console.log(
            `🕐 Server clock offset computed: ${this._serverClockOffsetMs.toFixed(0)}ms (client − server)`,
          );
        }
      },
    );

    // PlayStream event (new track/playlist)
    this.connection.on('PlayStream', (payload: PlayStreamPayload) => {
      console.log('📡 PlayStream event:', payload);
      this.eventHandlers.onPlayStream?.(payload);
    });

    // PlaybackStateChanged event (pause/resume/skip)
    this.connection.on(
      'PlaybackStateChanged',
      (payload: PlaybackStateChangedPayload) => {
        console.log('📡 PlaybackStateChanged event:', payload);
        this.eventHandlers.onPlaybackStateChanged?.(payload);
      },
    );

    // SpaceStateSync event (full state sync)
    this.connection.on('SpaceStateSync', (state: SpaceStateDto) => {
      console.log('📡 SpaceStateSync event:', state);
      this.eventHandlers.onSpaceStateSync?.(state.spaceId, state);
    });

    // Connection lifecycle events
    this.connection.onreconnecting(() => {
      console.log('🔄 SignalR reconnecting...');
      this.eventHandlers.onReconnecting?.();
    });

    this.connection.onreconnected(async () => {
      console.log('✅ SignalR reconnected');
      try {
        // Rejoin manager room after reconnect
        if (this.currentStoreId) {
          await this.joinStore(this.currentStoreId);
        }

        // Rejoin all active space groups after reconnect.
        const spaceIds = Array.from(this.joinedSpaceIds);
        await Promise.all(
          spaceIds.map(async (spaceId) => {
            try {
              await this.connection?.invoke('JoinSpaceAsync', spaceId);
              console.log('✅ Rejoined space group after reconnect:', spaceId);
            } catch (err) {
              console.error('❌ Failed to rejoin space group:', spaceId, err);
            }
          }),
        );
      } finally {
        this.eventHandlers.onReconnected?.();
      }
    });

    this.connection.onclose(() => {
      console.log('❌ SignalR connection closed');
      this.eventHandlers.onDisconnected?.();
    });
  }

  /**
   * Disconnect from StoreHub
   */
  public async disconnect(): Promise<void> {
    if (!this.connection) {
      return;
    }

    try {
      // Leave manager room
      await this.leaveStore();

      // Stop connection
      await this.connection.stop();

      console.log('👋 StoreHub disconnected gracefully');
    } catch (error) {
      console.error('❌ Error during disconnect:', error);
    } finally {
      this.connection = null;
      this.currentStoreId = null;
      this.joinedSpaceIds.clear();
      this.eventHandlers = {};
    }
  }
}

// Singleton instance
export const storeHubService = new StoreHubService();
